pkg/gatewayserver/io/basicstationlns/basicstationlns_test.go

```diff
@@ -0,0 +1,511 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package basicstationlns_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/smartystreets/assertions"
+	"go.thethings.network/lorawan-stack/pkg/basicstation"
+	"go.thethings.network/lorawan-stack/pkg/component"
+	"go.thethings.network/lorawan-stack/pkg/config"
+	"go.thethings.network/lorawan-stack/pkg/encoding/lorawan"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io"
+	. "go.thethings.network/lorawan-stack/pkg/gatewayserver/io/basicstationlns"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io/basicstationlns/messages"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io/mock"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
+	"go.thethings.network/lorawan-stack/pkg/util/test"
+	"go.thethings.network/lorawan-stack/pkg/util/test/assertions/should"
+)
+
+var (
+	registeredGatewayUID = "0101010101010101"
+	registeredGatewayID  = ttnpb.GatewayIdentifiers{GatewayID: "eui-" + registeredGatewayUID}
+	registeredGateway    = ttnpb.Gateway{GatewayIdentifiers: registeredGatewayID, FrequencyPlanID: "EU_863_870"}
+	registeredGatewayKey = "test-key"
+
+	discoveryEndPoint      = "ws://localhost:8100/api/v3/gs/io/basicstation/discover"
+	connectionRootEndPoint = "ws://localhost:8100/api/v3/gs/io/basicstation/traffic/"
+
+	testTrafficEndPoint = "ws://localhost:8100/api/v3/gs/io/basicstation/traffic/eui-0101010101010101"
+
+	timeout = 10 * test.Delay
+)
+
+func TestAuthentication(t *testing.T) {
+	// TODO: Test authentication. We're gonna provision authentication tokens, which may be API keys.
+	// https://github.com/TheThingsIndustries/lorawan-stack/issues/1413
+}
+
+func TestDiscover(t *testing.T) {
+	a := assertions.New(t)
+	ctx := log.NewContext(test.Context(), test.GetLogger(t))
+	ctx = newContextWithRightsFetcher(ctx)
+
+	c := component.MustNew(test.GetLogger(t), &component.Config{
+		ServiceBase: config.ServiceBase{
+			HTTP: config.HTTP{
+				Listen: ":8100",
```

Comment:
Please verify that we don't use this port in testing yet

Reply:
It's not used anywhere else as far as I can tell.